### PR TITLE
coqnative: do not load universe graph

### PR DESCRIPTION
**Pull request by SkySkimmer**
It is costly for no benefit.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
When reviewing remember that Library is a module internal to coqnative_bin and does not refer to the usual library/library.ml

**Review by gares — APPROVED** [review]

**Comment by SkySkimmer** [comment]
Bench got weirdly stuck in coq-performance-tests-lite (OLD), I ssh'd and did killall coqc

**Comment by maximedenes** [comment]
I'm sorry, I had to interrupt the bench as it was running on the only coq-bench machine I still have access to. The intermediate results were:
```
┌──────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬───────────────────────────┐
│                              │      user time [s]      │              CPU cycles               │           CPU instructions            │   max resident mem [KB]   │
│                              │                         │                                       │                                       │                           │
│         package_name         │   NEW      OLD    PDIFF │      NEW             OLD        PDIFF │      NEW             OLD        PDIFF │   NEW       OLD     PDIFF │
├──────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼───────────────────────────┤
│                  coq-coqutil │   53.33    55.39  -3.72 │   227534867788    236737369328  -3.89 │   360104679241    369288087095  -2.49 │   600332    600400  -0.01 │
│             coq-math-classes │  134.27   138.83  -3.28 │   579295564829    598216430020  -3.16 │   891550458926    917782467619  -2.86 │   513416    516524  -0.60 │
│                 coq-coqprime │   62.78    64.09  -2.04 │   271447928253    276100861617  -1.69 │   437784537035    443700214396  -1.33 │   765540    765292   0.03 │
│                  coq-bignums │   58.40    59.19  -1.33 │   244889347722    247051690000  -0.88 │   346287672519    348086053714  -0.52 │   717492    717436   0.01 │
│            coq-metacoq-pcuic │  832.77   842.79  -1.19 │  3615391441737   3660790251504  -1.24 │  5687139801392   5756350178741  -1.20 │  2196544   2196616  -0.00 │
│            coq-iris-examples │  548.28   554.56  -1.13 │  2385637191667   2413336452722  -1.15 │  3731418106948   3769139983415  -1.00 │  1151260   1151636  -0.03 │
│        coq-mathcomp-solvable │   92.30    92.99  -0.74 │   402752464670    405900411848  -0.78 │   636838485287    639003278426  -0.34 │   660480    660460   0.00 │
│         coq-mathcomp-algebra │   77.17    77.68  -0.66 │   335745176064    338003750484  -0.67 │   497009338169    499725862353  -0.54 │   570940    571704  -0.13 │
│           coq-mathcomp-field │  108.14   108.82  -0.62 │   472651886826    475414855587  -0.58 │   796366826936    798924162981  -0.32 │   653684    650572   0.48 │
│         coq-metacoq-template │  358.63   360.63  -0.55 │  1534062856653   1540942955988  -0.45 │  2652486596015   2663846976415  -0.43 │ 10012968  10012072   0.01 │
│      coq-metacoq-safechecker │  471.72   474.30  -0.54 │  2032812503267   2045669414503  -0.63 │  3345188474014   3363105687805  -0.53 │  9009292   9009188   0.00 │
│       coq-mathcomp-ssreflect │   35.53    35.72  -0.53 │   153329412311    154778068208  -0.94 │   218520698596    219855939599  -0.61 │   564104    564204  -0.02 │
│        coq-engine-bench-lite │  198.58   199.63  -0.53 │   833682621999    837389069405  -0.44 │  1574887674387   1581082996820  -0.39 │  1021016   1021028  -0.00 │
│       coq-mathcomp-odd-order │  479.42   481.95  -0.52 │  2098541517263   2109489174222  -0.52 │  3534320956033   3542584564095  -0.23 │   958340    959412  -0.11 │
│                 coq-bedrock2 │  424.68   426.76  -0.49 │  1853868167172   1864345730909  -0.56 │  3592003694467   3601079485820  -0.25 │  1022828   1022868  -0.00 │
│                     coq-core │  145.73   146.42  -0.47 │   460343648871    461044872828  -0.15 │   480786625714    481014650663  -0.05 │   286612    284760   0.65 │
│        coq-mathcomp-fingroup │   27.21    27.32  -0.40 │   118255348322    118973511529  -0.60 │   182266604105    183[1328](https://gitlab.com/coq/coq/-/jobs/3167583258#L1328)32359  -0.47 │   497660    498548  -0.18 │
│                 coq-compcert │  539.02   541.06  -0.38 │  2320487637601   2327325172608  -0.29 │  3883754161306   3880906977536   0.07 │  3837184   3837108   0.00 │
│                     coq-hott │  192.04   192.35  -0.16 │   800226409882    800766453250  -0.07 │  1167137536435   1167389801667  -0.02 │   631840    632272  -0.07 │
│       coq-mathcomp-character │   76.04    76.16  -0.16 │   332088958415    332918463522  -0.25 │   519490416476    520177774104  -0.13 │   739212    739260  -0.01 │
│                   coq-stdlib │  663.15   664.00  -0.13 │  2120222811907   2126513047200  -0.30 │  1925822017876   1925180350845   0.03 │   691552    689832   0.25 │
│ coq-fiat-crypto-with-bedrock │ 6148.65  6152.39  -0.06 │ 26806875254961  26830411086412  -0.09 │ 48362619449238  48372064030213  -0.02 │  4137340   4139152  -0.04 │
│                 coq-rewriter │  360.97   360.85   0.03 │  1572530018061   1571975709703   0.04 │  2462732488718   2460960248080   0.07 │  1259012   1259212  -0.02 │
│               coq-coquelicot │   40.08    39.93   0.38 │   172354967587    171644943110   0.41 │   243665167504    242781142507   0.36 │   748072    747404   0.09 │
│             coq-fiat-parsers │  350.97   348.56   0.69 │  1514273298823   1505783312566   0.56 │  2548067791105   2539016845223   0.36 │  3539632   3158092  12.08 │
│                coq-fiat-core │   58.89    58.47   0.72 │   244510583249    242880702909   0.67 │   358973709610    356582660913   0.67 │   486916    485036   0.39 │
└──────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴───────────────────────────┘
```

**Comment by SkySkimmer** [comment]
continued at https://gitlab.com/coq/coq/-/jobs/3171205223

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: kernel/safe_typing.mli [hunk 1]
```diff
@@ -225,7 +225,14 @@ val current_dirpath : safe_environment -> DirPath.t
 
 (** {6 Libraries : loading and saving compilation units } *)
 
-type compiled_library
+type library_info
+
+type compiled_library = private {
```
[the remark this answers is not in the document]
Comment (SkySkimmer) [reply kernel/safe_typing.mli]: In the end it seems we don't need a new API

**Comment by ppedrot** [comment]
@coqbot bench native

**Comment by ppedrot** [comment]
Is this really safe, though? We need the environment in the native compiler for three things, basically:
- Accessing inductive data for constructors (fine without universes)
- Expanding case nodes which is essentially the same use as above (fine without universes)
- Computation of the inductive type of fixpoint scrutinees and eta-expansion of cofixpoints.
Without universes I am not convinced that the third point will not raise anomalies. For instance, if we have definitional UIP set we definitely need the universe constraints to compute.

**Comment by SkySkimmer** [comment]
>Without universes I am not convinced that the third point will not raise anomalies. For instance, if we have definitional UIP set we definitely need the universe constraints to compute.

Yes
Not sure what to do about this

**Comment by ppedrot** [comment]
I think we should close this for now, but make the native compiler less demanding in its accesses to the the global environment.

**Pull request closed without merging**